framework/include/kernels/GenericKernelGrad.h

```diff
@@ -0,0 +1,53 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "KernelGrad.h"
+#include "ADKernelGrad.h"
+
+template <bool is_ad>
+class GenericKernelGrad : public KernelGrad
+{
+public:
+  static InputParameters validParams() { return KernelGrad::validParams(); };
+  GenericKernelGrad(const InputParameters & parameters) : KernelGrad(parameters) {}
+};
+
+template <>
+class GenericKernelGrad<true> : public ADKernelGrad
+{
+public:
+  static InputParameters validParams() { return ADKernelGrad::validParams(); };
+  GenericKernelGrad(const InputParameters & parameters) : ADKernelGrad(parameters) {}
+};
```

Comment:
here we also have a slightly more elegant pattern

```
template <bool is_ad>
using GenericKernelGradParent = std::conditional_t<is_ad, ADKernelGrad, KernelGrad>;
```

and then use `GenericKernelGradParent<is_ad>` as the parent class. No specialization needed.

Reply:
In fact you might as well just do 

```
template <bool is_ad>
using GenericKernelGrad = std::conditional_t<is_ad, ADKernelGrad, KernelGrad>;
```

No new classes required.

Reply:
i.e., the definition once of `usingGenericKernelGradMembers` to be inherited.